### PR TITLE
add /settings

### PR DESCRIPTION
**Pull request by MixailE**

### CODE REVIEW

**Comment by MixailE** [comment]
[Add a notification setting controller on ketl-notificaiton-backend](https://trello.com/c/RgTv4mXe/3832-add-a-notification-setting-controller-on-ketl-notificaiton-backend)

File: src/controllers/settings.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  async addSettings(
+    @Body({ required: true })
+    { allPostsEnabled, hotPostsEnabled, repliesEnabled, token }: Settings,
+    @Ctx() ctx: Context
+  ) {
+    try {
+      const tokenRecord = await TokenModel.findOne({
+        token,
+      })
+      if (!tokenRecord) return ctx.throw(badRequest(`Can't find token`))
```
Comment (T-Damer) [inline src/controllers/settings.ts:19]: Use plain `""`

File: src/controllers/settings.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+export default class SettingsController {
+  @Post('/')
+  async addSettings(
+    @Body({ required: true })
+    { allPostsEnabled, hotPostsEnabled, repliesEnabled, token }: Settings,
+    @Ctx() ctx: Context
+  ) {
+    try {
+      const tokenRecord = await TokenModel.findOne({
+        token,
```
Comment (T-Damer) [inline src/controllers/settings.ts:17]: We can remove `,` from here to make it one-line

File: src/controllers/settings.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      })
+      if (!tokenRecord) return ctx.throw(badRequest(`Can't find token`))
+
+      const settings = await SettingsModel.findOne({
+        token: tokenRecord,
+      })
+      if (settings) return { success: true }
```
Comment (T-Damer) [inline src/controllers/settings.ts:24]: Should we return updated settings?

File: src/controllers/settings.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          allPostsEnabled,
+          hotPostsEnabled,
+          repliesEnabled,
+        }
+      )
+
+      return { success: true }
+    } catch (e) {
+      console.error(e)
+      return ctx.throw(internal(`Can't save setting`))
```
Comment (T-Damer) [inline src/controllers/settings.ts:64]: I think `can't update settings` is better

File: src/controllers/settings.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    try {
+      const tokenRecord = await TokenModel.findOne({
+        token,
+      })
+      if (!tokenRecord) return ctx.throw(badRequest(`Can't find token`))
+
+      return findOneOrCreate(tokenRecord)
+    } catch (e) {
+      console.error(e)
+      return ctx.throw(internal(`Can't save token`))
```
Comment (T-Damer) [inline src/controllers/settings.ts:79]: `Can't find the token` or `Can't get the token`

File: src/controllers/settings.ts (hunk 3 above)
Comment (T-Damer) [inline src/controllers/settings.ts:24]: We should do: 
```TS
      await SettingsModel.updateOne(
        {
        token: tokenRecord,
        },
        { upsert: true } // Create a new document if it doesn't exist
      )
return {success : true}
```

Comment (T-Damer) [reply src/controllers/settings.ts:24]: Also, we may want to return actual `settings` to render them on the frontend

File: src/models/Settings.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  public token?: Ref<Token, string>
+  @prop({ default: true })
+  public allPostsEnabled?: boolean
+  @prop({ default: true })
+  public hotPostsEnabled?: boolean
+  @prop({ default: true })
+  public repliesEnabled?: boolean
+}
+
+export function findOne(token: Token) {
```
Comment (T-Damer) [inline src/models/Settings.ts:18]: `findSettingsByToken`

File: src/models/Token.ts [hunk 7]
```diff
@@ -4,6 +4,9 @@ import { getModelForClass, modelOptions, prop } from '@typegoose/typegoose'
   schemaOptions: { timestamps: true },
 })
 export class Token {
+  @prop()
+  public _id?: string
```
Comment (T-Damer) [inline src/models/Token.ts:8]: Why do we need this?

File: src/controllers/settings.ts (hunk 2 above)
Comment (MixailE) [reply src/controllers/settings.ts:17]: Nope, The linter don't allow that

File: src/controllers/settings.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import { Settings, SettingsModel, findOneOrCreate } from '@/models/Settings'
+import { TokenModel } from '@/models/Token'
+import { badRequest, internal } from '@hapi/boom'
+
+@Controller('/settings')
+export default class SettingsController {
+  @Post('/')
+  async addSettings(
+    @Body({ required: true })
+    { allPostsEnabled, hotPostsEnabled, repliesEnabled, token }: Settings,
```
Comment (T-Damer) [inline src/controllers/settings.ts:12]: Ruslan said that validator doesn't work when you destructure stuff like this, should we do:
```TS
body: Settings,

const { allPostsEnabled, hotPostsEnabled, repliesEnabled, token } = body
```
?

**Review by T-Damer — APPROVED** [review]

**Review by T-Damer — APPROVED** [review]

**Pull request merged**